File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.Queue;



public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));


    static int n, m;
    static char[][] map;
    static int[][] dist;
    static int[] dy = {1, -1, 0, 0};
    static int[] dx = {0, 0, -1, 1};

    static class Point {
        int y, x;
        Point(int y, int x){
            this.y = y;
            this.x = x;
        }
    }

    public static void main() throws IOException {

        String[] token = br.readLine().split(" ");
        n = Integer.valueOf(token[0]);
        m = Integer.valueOf(token[1]);
        map = new char[n][m];
        dist = new int[n][m];

        for (int i = 0; i < n; ++i){
            map[i] = br.readLine().toCharArray();
        }

        Queue<Point> q = new LinkedList<Point>();
        q.add(new Point(0, 0));
        dist[0][0] = 1;
        while(!q.isEmpty()){
            Point curr = q.poll();
            if (curr.y == n - 1 && curr.x == m - 1) break;

            for (int dir = 0; dir < 4; ++dir){
                int ny = curr.y + dy[dir];
                int nx = curr.x + dx[dir];
                if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
                if (map[ny][nx] == '0') continue;
                if (dist[ny][nx] != 0) continue;
                if (dist[ny][nx] >= dist[curr.y][curr.x] + 1) continue;
                q.add(new Point(ny, nx));
                dist[ny][nx] = dist[curr.y][curr.x] + 1;
            }
        }

        bw.write(dist[n - 1][m - 1] + "");

        br.close();
        bw.flush();
        bw.close();
    }


    public static void main(String[] args) throws IOException {
        Main.main();        
    }    
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.`out`))
lateinit var map: Array<CharArray>
lateinit var visited: Array<BooleanArray>
lateinit var trace: Array<IntArray>

fun bfs(n: Int, m: Int) {
    val dy = arrayOf(0, 0, -1, 1)
    val dx = arrayOf(-1, 1, 0, 0)
    val queue = LinkedList<Pair<Int, Int>>()

    queue.offer(Pair(0, 0))
    visited[0][0] = true
    trace[0][0] = 1

    while(queue.isNotEmpty()) {

        val (y, x) = queue.poll()
        if (y == n - 1 && x == m - 1) return
        for (dir in 0..<4) {
            val ny = y + dy[dir]
            val nx = x + dx[dir]

            if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue
            if (map[ny][nx] == '0') continue
            if (visited[ny][nx]) continue

            visited[ny][nx] = true
            trace[ny][nx] = trace[y][x] + 1
            queue.offer(Pair(ny, nx))
        }
    }
}

fun main() {
    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    map = Array(n){ CharArray(m) }
    trace = Array(n) { IntArray(m) }
    visited = Array(n) { BooleanArray(m) {false} }

    for (y in 0..<n){
        val line = br.readLine()
        map[y] = line.toCharArray()
    }

    bfs(n, m)

    bw.write("${trace[n-1][m-1]}")

    bw.flush()
    bw.close()
    br.close()
}


